File: webui/src/test/scala/com/taskadapter/reporting/ExportResultsFormatterTest.scala
package com.taskadapter.reporting

import com.taskadapter.connector.definition.TaskId
import com.taskadapter.web.uiapi.DecodedTaskError
import org.junit.runner.RunWith
import org.scalatest.{FunSpec, Matchers}
import org.scalatest.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
class ExportResultsFormatterTest extends FunSpec with Matchers {

  val id1 = TaskId(1, "key1")
  val id2 = TaskId(2, "key2")
  val id3 = TaskId(3, "key3")

  val item1 = new DecodedTaskError(id1, "error", "exception")
  val item1SameError = new DecodedTaskError(id2, "error", "exception")
  val item3 = new DecodedTaskError(id3, "error", "another exception")

  it("replaces duplicate exceptions") {
    ExportResultsFormatter.formatTaskErrors(Seq(item1, item1SameError)) shouldBe
      orig(id1) + line + same(id2)
  }

  it("replaces two duplicate exceptions") {
    ExportResultsFormatter.formatTaskErrors(Seq(
      new DecodedTaskError(id1, "error", "exception"),
      new DecodedTaskError(id2, "error", "exception"),
      new DecodedTaskError(id3, "error", "exception"))) shouldBe
      orig(id1) + line + same(id2) + line + same(id3)
  }

  it("leaves single element as is") {
    ExportResultsFormatter.formatTaskErrors(Seq(item1)) shouldBe orig(id1)
  }

  it("leaves different element in place") {
    ExportResultsFormatter.formatTaskErrors(Seq(item1, item3)) shouldBe
      orig(id1) + line + other(id3)
  }

  it("replaces duplicate and leaves different one in place") {
    ExportResultsFormatter.formatTaskErrors(Seq(item1, item1SameError, item3)) shouldBe
      orig(id1) + line + same(id2) + line + other(id3)
  }
  it("empty list gives empty string") {
    ExportResultsFormatter.formatTaskErrors(Seq()) shouldBe ""
  }

  def orig(id: TaskId): String = {
    s"$id - error - exception"
  }

  def same(id: TaskId): String = {
    s"$id - error - same as previous"
  }

  def other(id: TaskId): String = {
    s"$id - error - another exception"
  }

  private def line = System.lineSeparator()
}

File: webui/src/test/java/com/taskadapter/reporting/ExportResultsFormatterTest.java
package com.taskadapter.reporting;

import com.taskadapter.connector.definition.TaskId;
import com.taskadapter.web.uiapi.DecodedTaskError;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;

import static com.taskadapter.reporting.ExportResultsFormatter.formatTaskErrors;
import static org.assertj.core.api.Assertions.assertThat;

public class ExportResultsFormatterTest {
    private static final TaskId id1 = new TaskId(1L, "key1");
    private static final TaskId id2 = new TaskId(2L, "key2");
    private static final TaskId id3 = new TaskId(3L, "key3");

    private static final DecodedTaskError item1 = new DecodedTaskError(id1, "error", "exception");
    private static final DecodedTaskError item1SameError = new DecodedTaskError(id2, "error", "exception");
    private static final DecodedTaskError item3 = new DecodedTaskError(id3, "error", "another exception");

    private static final String line = System.lineSeparator();

    @Test
    public void replacesDuplicateExceptions() {
        assertThat(formatTaskErrors(java.util.List.of(item1, item1SameError)))
                .isEqualTo(orig(id1) + line + same(id2));
    }

    @Test
    public void replacesTwoDuplicateExceptions() {
        assertThat(
                formatTaskErrors(List.of(
                        new DecodedTaskError(id1, "error", "exception"),
                        new DecodedTaskError(id2, "error", "exception"),
                        new DecodedTaskError(id3, "error", "exception"))))
                .isEqualTo(
                        orig(id1) + line + same(id2) + line + same(id3));
    }

    @Test
    public void leavesSingleElementAsIs() {
        assertThat(formatTaskErrors(Arrays.asList(item1)))
                .isEqualTo(orig(id1));
    }

    @Test
    public void leavesDifferentElementInPlace() {
        assertThat(formatTaskErrors(List.of(item1, item3)))
                .isEqualTo(orig(id1) + line + other(id3));
    }

    @Test
    public void replacesDuplicateAndLeavesDifferentOneInPlace() {
        assertThat(formatTaskErrors(List.of(item1, item1SameError, item3)))
                .isEqualTo(orig(id1) + line + same(id2) + line + other(id3));
    }

    @Test
    public void emptyListGivesEmptyString() {
        assertThat(formatTaskErrors(List.of())).isEqualTo("");
    }

    private static String orig(TaskId id) {
        return id + " - error - exception";
    }

    private static String same(TaskId id) {
        return id + " - error - same as previous";
    }

    private static String other(TaskId id) {
        return id + " - error - another exception";
    }
}
